### ISSUE TITLE
persistent_volumes_test integration test flaky on shipable

### ISSUE THREAD

**Issue by jszczepkowski** [kubernetes/kubernetes#14107]
Test persistent_volumes_test.go is flaky on shipable. It fails on go 1.3 due to data race:

WARNING: DATA RACE
Read by goroutine 190:
  github.com/coreos/go-etcd/etcd.(*Client).getCancelable()
      /root/workspace/src/k8s.io/kubernetes/Godeps/_workspace/src/github.com/coreos/go-etcd/etcd/requests.go:41 +0xcc
  github.com/coreos/go-etcd/etcd.(*Client).get()
      /root/workspace/src/k8s.io/kubernetes/Godeps/_workspace/src/github.com/coreos/go-etcd/etcd/requests.go:62 +0x78
  github.com/coreos/go-etcd/etcd.(*Client).RawGet()
  k8s.io/kubernetes/pkg/registry/namespace.(*storage).GetNamespace()
      /root/workspace/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/registry/namespace/registry.go:67 +0xb5
  k8s.io/kubernetes/pkg/master.(*Controller).CreateNamespaceIfNeeded()
      /root/workspace/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/master/controller.go:126 +0xd3
  k8s.io/kubernetes/pkg/master.(*Controller).UpdateKubernetesService()
      /root/workspace/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/master/controller.go:109 +0x63
  k8s.io/kubernetes/pkg/master.func·001()
      /root/workspace/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/master/controller.go:97 +0x54
  k8s.io/kubernetes/pkg/util.func·002()
      /root/workspace/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/util/util.go:115 +0x6a
  k8s.io/kubernetes/pkg/util.Until()
      /root/workspace/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/registry/generic/etcd/etcd.go:443 +0x123
  k8s.io/kubernetes/pkg/registry/persistentvolume/etcd.(*REST).Watch()
      :9 +0xfa
  k8s.io/kubernetes/pkg/apiserver.func·024()
      /root/workspace/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/apiserver/resthandler.go:253 +0xe81
  github.com/emicklei/go-restful.func·009()
      /root/workspace/src/k8s.io/kubernetes/Godeps/_workspace/src/github.com/emicklei/go-restful/container.go:244 +0xa7
  github.com/emicklei/go-restful.(*FilterChain).ProcessFilter()
      /root/workspace/src/k8s.io/kubernetes/Godeps/_workspace/src/github.com/emicklei/go-restful/filter.go:21 +0x183
  k8s.io/kubernetes/pkg/apiserver.func·003()
      /root/workspace/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/apiserver/apiserver.go:61 +0x9c
  github.com/emicklei/go-restful.(*FilterChain).ProcessFilter()
      /root/workspace/src/k8s.io/kubernetes/Godeps/_workspace/src/github.com/emicklei/go-restful/filter.go:19 +0x13c
  github.com/emicklei/go-restful.(*Container).dispatch()
      /root/workspace/src/k8s.io/kubernetes/Godeps/_workspace/src/github.com/emicklei/go-restful/container.go:246 +0x129d
  github.com/emicklei/go-restful.*Container.(github.com/emicklei/go-restful.dispatch)·fm()
      /root/workspace/src/k8s.io/kubernetes/Godeps/_workspace/src/github.com/emicklei/go-restful/container.go:95 +0x51
  net/http.HandlerFunc.ServeHTTP()
      /root/.gvm/gos/go1.3/src/pkg/net/http/server.go:1235 +0x4d
  net/http.(*ServeMux).ServeHTTP()
      /root/.gvm/gos/go1.3/src/pkg/net/http/server.go:1511 +0x21c
  k8s.io/kubernetes/pkg/apiserver.func·014()
      /root/workspace/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/util/util.go:116 +0xbb
  k8s.io/kubernetes/pkg/master.(*Controller).RunKubernetesService()
      /root/workspace/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/master/controller.go:100 +0xa1
  k8s.io/kubernetes/pkg/master.*Controller.RunKubernetesService·fm()
      /root/workspace/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/master/controller.go:90 +0x3d

Previous write by goroutine 253:
  github.com/coreos/go-etcd/etcd.(*Cluster).failure()
      /root/workspace/src/k8s.io/kubernetes/Godeps/_workspace/src/github.com/coreos/go-etcd/etcd/cluster.go:28 +0x69
  github.com/coreos/go-etcd/etcd.(*Client).SendRequest()
      /root/workspace/src/k8s.io/kubernetes/Godeps/_workspace/src/github.com/coreos/go-etcd/etcd/requests.go:264 +0x17a6
      /root/workspace/src/k8s.io/kubernetes/Godeps/_workspace/src/github.com/coreos/go-etcd/etcd/get.go:31 +0x28e
  github.com/coreos/go-etcd/etcd.(*Client).Get()
      /root/workspace/src/k8s.io/kubernetes/Godeps/_workspace/src/github.com/coreos/go-etcd/etcd/get.go:11 +0x79
  k8s.io/kubernetes/pkg/storage/etcd.(*etcdHelper).bodyAndExtractObj()
      /root/workspace/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/storage/etcd/etcd_helper.go:205 +0xf3
  k8s.io/kubernetes/pkg/storage/etcd.(*etcdHelper).Get()
      /root/workspace/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/storage/etcd/etcd_helper.go:197 +0xc4
  k8s.io/kubernetes/pkg/registry/generic/etcd.(*Etcd).Get()
      /root/workspace/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/registry/generic/etcd/etcd.go:332 +0x2d9
  k8s.io/kubernetes/pkg/registry/namespace/etcd.(*REST).Get()
      :2 +0xaf
  github.com/coreos/go-etcd/etcd.(*Client).getCancelable()
      /root/workspace/src/k8s.io/kubernetes/Godeps/_workspace/src/github.com/coreos/go-etcd/etcd/requests.go:51 +0x4a5
  github.com/coreos/go-etcd/etcd.(*Client).watchOnce()
      /root/workspace/src/k8s.io/kubernetes/Godeps/_workspace/src/github.com/coreos/go-etcd/etcd/watch.go:96 +0x20b
  github.com/coreos/go-etcd/etcd.(*Client).Watch()
      /root/workspace/src/k8s.io/kubernetes/Godeps/_workspace/src/github.com/coreos/go-etcd/etcd/watch.go:39 +0x32f
  k8s.io/kubernetes/pkg/storage/etcd.(*etcdWatcher).etcdWatch()
      /root/workspace/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/storage/etcd/etcd_watcher.go:131 +0x24e

Goroutine 190 (running) created at:
  k8s.io/kubernetes/pkg/util.(*Runner).Start()
      /root/workspace/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/apiserver/handlers.go:399 +0x153
  net/http.HandlerFunc.ServeHTTP()
      /root/.gvm/gos/go1.3/src/pkg/net/http/server.go:1235 +0x4d
  k8s.io/kubernetes/pkg/api.func·014()
      /root/workspace/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/api/requestcontext.go:101 +0x11c
  net/http.HandlerFunc.ServeHTTP()
      /root/.gvm/gos/go1.3/src/pkg/net/http/server.go:1235 +0x4d
  k8s.io/kubernetes/test/integration.func·001()
      /root/workspace/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/integration/utils.go:87 +0x9f
  net/http.HandlerFunc.ServeHTTP()
      /root/.gvm/gos/go1.3/src/pkg/net/http/server.go:1235 +0x4d
      /root/workspace/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/util/runner.go:45 +0x1f6
  k8s.io/kubernetes/pkg/master.(*Controller).Start()
      /root/workspace/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/master/controller.go:91 +0x8ed
  k8s.io/kubernetes/pkg/master.(*Controller).Start()
      /root/workspace/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/master/controller.go:86 +0x5f2
  k8s.io/kubernetes/pkg/master.(*Master).init()
      /root/workspace/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/master/master.go:656 +0x572f
  k8s.io/kubernetes/pkg/master.New()
      /root/workspace/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/master/master.go:394 +0xc21
  k8s.io/kubernetes/test/integration.runAMaster()
      /root/workspace/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/integration/utils.go:84 +0x57a
  k8s.io/kubernetes/test/integration.TestPersistentVolumeRecycler()
      /root/workspace/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/integration/persistent_volumes_test.go:212 +0x45
  testing.tRunner()
      /root/.gvm/gos/go1.3/src/pkg/testing/testing.go:422 +0x10f

Goroutine 253 (running) created at:
  k8s.io/kubernetes/pkg/storage/etcd.(*etcdHelper).WatchList()
      /root/workspace/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/storage/etcd/etcd_helper.go:190 +0x211
  k8s.io/kubernetes/pkg/registry/generic/etcd.(*Etcd).WatchPredicate()
      /root/workspace/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/registry/generic/etcd/etcd.go:464 +0x40a
  k8s.io/kubernetes/pkg/registry/generic/etcd.(*Etcd).Watch()
  net/http/httptest.(*waitGroupHandler).ServeHTTP()
      /root/.gvm/gos/go1.3/src/pkg/net/http/httptest/server.go:200 +0xf4
  net/http.serverHandler.ServeHTTP()
      /root/.gvm/gos/go1.3/src/pkg/net/http/server.go:1673 +0x1fc
  net/http.(*conn).serve()
      /root/.gvm/gos/go1.3/src/pkg/net/http/server.go:1174 +0xf9e

**Comment by jszczepkowski** [comment]
I experienced this problem two times so far today. Logs can be found at (build 26377):

https://app.shippable.com/builds/55faafcae5b1f60c00919d3f

**Comment by jszczepkowski** [comment]
@markturansky 
It seems that there is a problem with TestPersistentVolumeRecycler, which you added quite recently.

**Comment by markturansky** [comment]
I've noticed the same with Go 1.3.  I will attempt to recreate locally and debug it.

**Comment by jszczepkowski** [comment]
Similar problem occurs in TestPersistentVolumeDeleter:
https://app.shippable.com/builds/55fbfc0f9b216d0d00d23883

**Comment by smarterclayton** [comment]
The data race is in the etcd client - so mark, are you using a registry and a fake etcd client and reading from one of those objects in a different goroutine?

**Comment by smarterclayton** [comment]
```
/root/workspace/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/master/master.go:394 +0xc21
k8s.io/kubernetes/test/integration.runAMaster()
/root/workspace/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/integration/utils.go:84 +0x57a
k8s.io/kubernetes/test/integration.TestPersistentVolumeRecycler()
/root/workspace/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/integration/persistent_volumes_test.go:212 +0x45
```

The integration test is starting a master?  That's not safe from our integration structure today unless the integration test harness runs each test in a unique process.  What other code is using a master in integration tests?

**Comment by markturansky** [comment]
I think most of them work that way.  I copied the format for my test from others.

**Comment by smarterclayton** [comment]
What happens when you do that is that goroutines are running in the
background from previous executions.  Are you reusing an etcd client from
them?

On Fri, Sep 18, 2015 at 12:38 PM, Mark Turansky <notifications@github.com>
wrote:

> I think most of them work that way. I copied the format for my test from
> others.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/issues/14107#issuecomment-141503003>
> .
>



-- 
Clayton Coleman | Lead Engineer, OpenShift

**Comment by markturansky** [comment]
No.  

Each TestWhateverFoo that uses a master all follow this pattern:

```
	_, s := runAMaster(t)
	defer s.Close()
	deleteAllEtcdKeys()
	client := client.NewOrDie(&client.Config{Host: s.URL, Version: testapi.Default.Version()})
```

But if each TestWhateverFoo is running concurrently, then there are many masters?

**Comment by markturansky** [comment]
I added an integration test for the Recycler and Deleter.  I already had one testing the binder.

That's 3 concurrently running integration tests (if they run in parallel) that have a binder running within.  The other two have a binder and a recycler controller running within.

**Comment by smarterclayton** [comment]
Is someone changing the etcd in the background after each integration test?

**Comment by markturansky** [comment]
After each test?  No.  All modifications happening with TestWhateverFoo func.

**Comment by markturansky** [comment]
I have an open PR that is related to my integration tests.  I gave each of the 3 callers their own instance of client.

If the client is not internally synchronized and I fix the flake with more client instances, then kube controller manager is inherently not thread safe.  All controllers, to my knowledge, share the same client.  

Perhaps this is not an issue for most controllers, but I have a few that all work on the same type of API object.  This might be the data race we're seeing.

**Comment by smarterclayton** [comment]
I would expect every test case in integration that uses runAMaster to hit this.  etcd client is not thread safe, and while the thread safety problem doesn't cause failures, it means race detector will catch it.

Solutions:

1. Switch to thread safe etcd client (@tstclair was working on this elsewhere)
2. Disable race detection on some integration tests

That's it.

**Comment by quinton-hoole** [comment]
@markturansky ping?  This issue is holding up all merges, as integration tests are not passing consistently.  Are you still actively working on this?

**Comment by markturansky** [comment]
@quinton-hoole yes.  attempting to reproduce.

**Comment by saad-ali** [comment]
Mark, here is another example of this failure, if that helps:

https://app.shippable.com/builds/56006ef0b88a390c009e4e52

```
    === RUN TestPersistentVolumeClaimBinder

    W0921 21:15:59.609410   17011 master.go:261] Network range for service cluster IPs is unspecified. Defaulting to 10.0.0.0/24.

    I0921 21:15:59.609570   17011 master.go:287] Node port range unspecified. Defaulting to 30000-32767.

    I0921 21:15:59.610244   17011 master.go:309] Will report 172.17.4.148 as public IP address.

    W0921 21:15:59.802258   17011 controller.go:212] Resetting endpoints for master service "kubernetes" to &{{ } {kubernetes  default    0 0001-01-01 00:00:00 +0000 UTC <nil> <nil> map[] map[]} [{[{172.17.4.148 <nil>}] [] [{ 6443 TCP}]}]}

    E0921 21:15:59.806517   17011 repair.go:54] unable to persist the updated port allocations: 100: Key not found (/kubernetes.io) [160]

    E0921 21:16:00.004954   17011 persistentvolume_claim_binder_controller.go:132] PVClaimBinder could not add claim claim03: A volume match does not exist for persistent claim: claim03

    --- PASS: TestPersistentVolumeClaimBinder (0.56 seconds)
...
    ==================

    WARNING: DATA RACE

    Read by goroutine 68:

      github.com/coreos/go-etcd/etcd.(*Client).getCancelable()

          /root/workspace/src/k8s.io/kubernetes/Godeps/_workspace/src/github.com/coreos/go-etcd/etcd/requests.go:41 +0xcc

      github.com/coreos/go-etcd/etcd.(*Client).get()

          /root/workspace/src/k8s.io/kubernetes/Godeps/_workspace/src/github.com/coreos/go-etcd/etcd/requests.go:62 +0x78

      github.com/coreos/go-etcd/etcd.(*Client).RawGet()

          /root/workspace/src/k8s.io/kubernetes/Godeps/_workspace/src/github.com/coreos/go-etcd/etcd/get.go:31 +0x28e

      github.com/coreos/go-etcd/etcd.(*Client).Get()

          /root/workspace/src/k8s.io/kubernetes/Godeps/_workspace/src/github.com/coreos/go-etcd/etcd/get.go:11 +0x79

      k8s.io/kubernetes/pkg/storage/etcd.(*etcdHelper).bodyAndExtractObj()

          /root/workspace/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/storage/etcd/etcd_helper.go:205 +0xf3

      k8s.io/kubernetes/pkg/storage/etcd.(*etcdHelper).Get()

          /root/workspace/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/storage/etcd/etcd_helper.go:197 +0xc4

      k8s.io/kubernetes/pkg/registry/generic/etcd.(*Etcd).Get()

          /root/workspace/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/registry/generic/etcd/etcd.go:333 +0x2d9

      k8s.io/kubernetes/pkg/registry/namespace/etcd.(*REST).Get()

          <autogenerated>:2 +0xaf

      k8s.io/kubernetes/pkg/registry/namespace.(*storage).GetNamespace()

          /root/workspace/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/registry/namespace/registry.go:67 +0xb5

      k8s.io/kubernetes/pkg/master.(*Controller).CreateNamespaceIfNeeded()

          /root/workspace/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/master/controller.go:126 +0xd3

      k8s.io/kubernetes/pkg/master.(*Controller).UpdateKubernetesService()

          /root/workspace/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/master/controller.go:109 +0x63

      k8s.io/kubernetes/pkg/master.func·001()

          /root/workspace/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/master/controller.go:97 +0x54

      k8s.io/kubernetes/pkg/util.func·002()

          /root/workspace/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/util/util.go:116 +0x6a

      k8s.io/kubernetes/pkg/util.Until()

          /root/workspace/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/util/util.go:117 +0xbb

      k8s.io/kubernetes/pkg/util.Until()

          /root/workspace/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/util/util.go:117 +0xbb

      k8s.io/kubernetes/pkg/master.(*Controller).RunKubernetesService()

          /root/workspace/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/master/controller.go:100 +0xa1

      k8s.io/kubernetes/pkg/master.*Controller.RunKubernetesService·fm()

          /root/workspace/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/master/controller.go:90 +0x3d

    

    Previous write by goroutine 207:

      [failed to restore the stack]

    

    Goroutine 68 (running) created at:

      k8s.io/kubernetes/pkg/util.(*Runner).Start()

          /root/workspace/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/util/runner.go:45 +0x1f6

      k8s.io/kubernetes/pkg/master.(*Controller).Start()

          /root/workspace/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/master/controller.go:91 +0x8ed

      k8s.io/kubernetes/pkg/master.(*Master).init()

          /root/workspace/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/master/master.go:656 +0x572f

      k8s.io/kubernetes/pkg/master.New()

          /root/workspace/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/master/master.go:394 +0xc21

      k8s.io/kubernetes/test/integration.runAMaster()

          /root/workspace/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/integration/utils.go:84 +0x57a

      k8s.io/kubernetes/test/integration.TestPersistentVolumeClaimBinder()

          /root/workspace/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/integration/persistent_volumes_test.go:41 +0x45

      testing.tRunner()

          /root/.gvm/gos/go1.3/src/pkg/testing/testing.go:422 +0x10f

    

    Goroutine 207 (running) created at:

      k8s.io/kubernetes/pkg/storage/etcd.(*etcdHelper).WatchList()

          /root/workspace/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/storage/etcd/etcd_helper.go:190 +0x211

      k8s.io/kubernetes/pkg/registry/generic/etcd.(*Etcd).WatchPredicate()

          /root/workspace/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/registry/generic/etcd/etcd.go:465 +0x40a

      k8s.io/kubernetes/pkg/registry/generic/etcd.(*Etcd).Watch()

          /root/workspace/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/registry/generic/etcd/etcd.go:444 +0x123

      k8s.io/kubernetes/pkg/registry/persistentvolumeclaim/etcd.(*REST).Watch()

          <autogenerated>:9 +0xfa

      k8s.io/kubernetes/pkg/apiserver.func·024()

          /root/workspace/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/apiserver/resthandler.go:254 +0xe81

      github.com/emicklei/go-restful.func·009()

          /root/workspace/src/k8s.io/kubernetes/Godeps/_workspace/src/github.com/emicklei/go-restful/container.go:244 +0xa7

      github.com/emicklei/go-restful.(*FilterChain).ProcessFilter()

          /root/workspace/src/k8s.io/kubernetes/Godeps/_workspace/src/github.com/emicklei/go-restful/filter.go:21 +0x183

      k8s.io/kubernetes/pkg/apiserver.func·003()

          /root/workspace/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/apiserver/apiserver.go:61 +0x9c

      github.com/emicklei/go-restful.(*FilterChain).ProcessFilter()

          /root/workspace/src/k8s.io/kubernetes/Godeps/_workspace/src/github.com/emicklei/go-restful/filter.go:19 +0x13c

      github.com/emicklei/go-restful.(*Container).dispatch()

          /root/workspace/src/k8s.io/kubernetes/Godeps/_workspace/src/github.com/emicklei/go-restful/container.go:246 +0x129d

      github.com/emicklei/go-restful.*Container.(github.com/emicklei/go-restful.dispatch)·fm()

          /root/workspace/src/k8s.io/kubernetes/Godeps/_workspace/src/github.com/emicklei/go-restful/container.go:95 +0x51

      net/http.HandlerFunc.ServeHTTP()

          /root/.gvm/gos/go1.3/src/pkg/net/http/server.go:1235 +0x4d

      net/http.(*ServeMux).ServeHTTP()

          /root/.gvm/gos/go1.3/src/pkg/net/http/server.go:1511 +0x21c

      k8s.io/kubernetes/pkg/apiserver.func·014()

          /root/workspace/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/apiserver/handlers.go:399 +0x153

      net/http.HandlerFunc.ServeHTTP()

          /root/.gvm/gos/go1.3/src/pkg/net/http/server.go:1235 +0x4d

      k8s.io/kubernetes/pkg/api.func·014()

          /root/workspace/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/api/requestcontext.go:101 +0x11c

      net/http.HandlerFunc.ServeHTTP()

          /root/.gvm/gos/go1.3/src/pkg/net/http/server.go:1235 +0x4d

      k8s.io/kubernetes/test/integration.func·001()

          /root/workspace/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/integration/utils.go:87 +0x9f

      net/http.HandlerFunc.ServeHTTP()

          /root/.gvm/gos/go1.3/src/pkg/net/http/server.go:1235 +0x4d

      net/http/httptest.(*waitGroupHandler).ServeHTTP()

          /root/.gvm/gos/go1.3/src/pkg/net/http/httptest/server.go:200 +0xf4

      net/http.serverHandler.ServeHTTP()

          /root/.gvm/gos/go1.3/src/pkg/net/http/server.go:1673 +0x1fc

      net/http.(*conn).serve()

          /root/.gvm/gos/go1.3/src/pkg/net/http/server.go:1174 +0xf9e

    ==================
...
    Found 1 data race(s)

    exit status 66

    FAIL	k8s.io/kubernetes/test/integration	109.585s
```

Note that this is a different issue then #14284.

**Comment by dchen1107** [comment]
Another failure: https://app.shippable.com/builds/5600d606f4951e0c00e841a1

**Comment by smarterclayton** [comment]
This is blocked by the go-etcd client - all we can do is disable race detection on integration.

**Comment by saad-ali** [comment]
Closed with #14356
